src/device_trezor/device_trezor.cpp

```diff
@@ -213,16 +313,144 @@ namespace trezor {
         memcpy(sig.r.data, buff + 64, 32);
         ski.push_back(std::make_pair(ki, sig));
       }
+#undef EVENT_PROGRESS
+    }
+
+    bool device_trezor::is_live_refresh_supported() const
+    {
+      require_initialized();
+      return get_version() > pack_version(2, 0, 10);
+    }
+
+    bool device_trezor::is_live_refresh_enabled() const
+    {
+      return is_live_refresh_supported() && (mode == NONE || mode == TRANSACTION_PARSE) && m_live_refresh_enabled;
+    }
+
+    bool device_trezor::has_ki_live_refresh() const
+    {
+      try{
+        return is_live_refresh_enabled();
+      } catch(const std::exception & e){
+        MERROR("Could not detect if live refresh is enabled: " << e.what());
+      }
+      return false;
+    }
+
+    void device_trezor::live_refresh_start()
+    {
+      AUTO_LOCK_CMD();
+      require_connected();
+      device_state_reset_unsafe();
+      require_initialized();
+
+      auto req = std::make_shared<messages::monero::MoneroLiveRefreshStartRequest>();
+      this->set_msg_addr<messages::monero::MoneroLiveRefreshStartRequest>(req.get());
+      this->client_exchange<messages::monero::MoneroLiveRefreshStartAck>(req);
+      m_live_refresh_in_progress = true;
+    }
+
+    void device_trezor::live_refresh(
+        const ::crypto::secret_key & view_key_priv,
+        const crypto::public_key& out_key,
+        const crypto::key_derivation& recv_derivation,
+        size_t real_output_index,
+        const cryptonote::subaddress_index& received_index,
+        cryptonote::keypair& in_ephemeral,
+        crypto::key_image& ki
+    )
+    {
+      require_connected();
+      if (!m_live_refresh_in_progress)
+      {
+        live_refresh_start();
+      }
+
+      m_last_live_refresh_time = std::chrono::system_clock::now();
```

Comment:
You probably want to set this before calling live_refresh_start, or you might time out early if unlucky.

Reply:
I've set it after the `live_refresh_start` because it sets `m_live_refresh_in_progress` which enables the thread auto-shutdown. The `live_refresh_start` is user-confirmed action so the idea was to set the `m_last_live_refresh_time` after this.

Reply:
Nevermind, I confused m_live_refresh_running and m_live_refresh_in_progress.